### PR TITLE
ardour: add ardour 6 and make it the default

### PR DESCRIPTION
**Pull request by magnetophon**
We leave Ardour 5 in, since Ardour 6 has a new latency compensation
algorithm, so your mixes might end up sounding different in the new version.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/audio/ardour/ardour5.nix [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+, libgnomecanvas
+, libgnomecanvasmm
+, liblo
+, libogg
+, librdf_raptor
+, librdf_rasqal
+, libsamplerate
+, libsigcxx
+, libsndfile
+, libusb
```
Comment (cole-h) [inline pkgs/applications/audio/ardour/ardour5.nix:25]: `libusb` is an alias (which we don't allow in packages). It is aliased to `libusb1`, so that might fix borg's discontent (at least about using aliases):

```suggestion
, libusb1
```

Comment (magnetophon) [reply pkgs/applications/audio/ardour/ardour5.nix]: ok, done. Thanks!

File: pkgs/applications/audio/ardour/default.nix [hunk 2]
```diff
@@ -127,21 +143,22 @@ in stdenv.mkDerivation rec {
   # be available. Since this is an unzipped tarball fetched from github we
   # have to do that ourself.
   postPatch = ''
-    printf '#include "libs/ardour/ardour/revision.h"\nnamespace ARDOUR { const char* revision = \"${tag}-${builtins.substring 0 8 src.rev}\"; }\n' > libs/ardour/revision.cc
+    printf '#include "libs/ardour/ardour/revision.h"\nnamespace ARDOUR { const char* revision = "6.0"; const char* date = "2020-05-23"; }\n' > libs/ardour/revision.cc
+    sed 's|/usr/include/libintl.h|${glibc.dev}/include/libintl.h|' -i wscript
     patchShebangs ./tools/
+    sed "382 s/mingw/x86_64/" -i libs/ardour/wscript
```
Comment (magnetophon) [inline pkgs/applications/audio/ardour/default.nix:149]: This is a bit of a hack to get it to compile.
Open to suggestions!

Comment (cole-h) [reply pkgs/applications/audio/ardour/default.nix:149]: A better solution would be an actual patch. This is too brittle, because (IIRC) a failed `sed` won't halt the build. A failed patch will.

File: pkgs/applications/audio/ardour/ardour5.nix [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+    vamp-plugin-sdk
+    libarchive
+  ];
+
+  wafConfigureFlags = [
+    "--optimize"
+    "--docs"
+    "--use-external-libs"
+    "--freedesktop"
+    "--with-backends=jack,alsa,dummy"
```
Comment (x42) [inline pkgs/applications/audio/ardour/ardour5.nix:121]: This option is no longer needed, by default Ardour picks all available
(which is now equivalent to `--with-backends=jack,alsa,dummy,pulseaudio"`

File: pkgs/applications/audio/ardour/ardour5.nix [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+  postPatch = ''
+    printf '#include "libs/ardour/ardour/revision.h"\nnamespace ARDOUR { const char* revision = \"${tag}-${builtins.substring 0 8 src.rev}\"; }\n' > libs/ardour/revision.cc
+    patchShebangs ./tools/
+  '';
+
+  postInstall = ''
+    # wscript does not install these for some reason
+    install -vDm 644 "build/gtk2_ardour/ardour.xml" \
+      -t "$out/share/mime/packages"
+    install -vDm 644 "build/gtk2_ardour/ardour5.desktop" \
```
Comment (x42) [inline pkgs/applications/audio/ardour/ardour5.nix:138]: ardour6.desktop

File: pkgs/applications/audio/ardour/default.nix (hunk 2 above)
Comment (magnetophon) [reply pkgs/applications/audio/ardour/default.nix:149]: To make a proper patch I need to know what to change this code to.
I tried removing the if statement altogether, but got errors about unexpected indentation on line 383.
Tried to remove the indentation but kept getting the same error.
ATM I'm just changing the condition, but that only works when building for x86_64.

**Comment by x42** [comment]
looks like you've missed libpulse-dev and libwebsockes-dev packages as dependency

File: pkgs/applications/audio/ardour/default.nix (hunk 2 above)
Comment (cole-h) [reply pkgs/applications/audio/ardour/default.nix:149]: Well, you're already changing that part from `mingw` to `x86_64` -- why not a patch that does the same? This way, git will be able to automatically resolve slight location changes (e.g. added line here or there), but will fail if the patch doesn't apply (e.g. that line gets removed entirely).

Comment (x42) [reply pkgs/applications/audio/ardour/default.nix:149]: This could go upstream, but I'd first like to understand why it's needed for NixOS only.
other distros with the  GNU assembler  2.31.1 don't have this issue.

File: pkgs/applications/audio/ardour/ardour5.nix (hunk 4 above)
Comment (magnetophon) [reply pkgs/applications/audio/ardour/ardour5.nix:138]: That's the A5 pkg you're looking at.

File: pkgs/applications/audio/ardour/ardour5.nix [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+
+  # ardour's wscript has a "tarball" target but that required the git revision
+  # be available. Since this is an unzipped tarball fetched from github we
+  # have to do that ourself.
+  postPatch = ''
+    printf '#include "libs/ardour/ardour/revision.h"\nnamespace ARDOUR { const char* revision = \"${tag}-${builtins.substring 0 8 src.rev}\"; }\n' > libs/ardour/revision.cc
+    patchShebangs ./tools/
+  '';
+
+  postInstall = ''
```
Comment (magnetophon) [inline pkgs/applications/audio/ardour/ardour5.nix:134]: @x42 Should the wscript do this in theory?  
If yes, any idea why it doesn't here?
If no, why not?

File: pkgs/applications/audio/ardour/default.nix (hunk 2 above)
Comment (magnetophon) [reply pkgs/applications/audio/ardour/default.nix:149]: @cole-h That "fix" only works when building for x86_64.  Ardour should run on many platforms.

Comment (cole-h) [reply pkgs/applications/audio/ardour/default.nix:149]: A failed `sed` does not halt the build. I changed this to `sed "382 s/fdjsaklfjldska/x86_64/" -i libs/ardour/wscript` and it continued building just fine (at least, until it errored out due to what this is trying to fix).

`echo asdf | sed "s/fdsa/jkl;" && echo $?` -> `asdf` and `0`.

But yes, tracking down why this is necessary is a better idea.

File: pkgs/applications/audio/ardour/ardour5.nix (hunk 3 above)
Comment (magnetophon) [reply pkgs/applications/audio/ardour/ardour5.nix:121]: Thanks, done.

**Comment by magnetophon** [comment]
> looks like you've missed libpulse-dev and libwebsockes-dev packages as dependency

pulseaudio is in there,and it claims to build with it. I don't have pulse installed,so have not tested. 

seems that in order to build with libwebsockets, you also need libuv (assuming that's what provides uv.h), but it only fails halfway trough the build.
Both are in now, thanks.

File: pkgs/top-level/all-packages.nix [hunk 6]
```diff
@@ -18634,6 +18634,8 @@ in
 
   ardour = callPackage ../applications/audio/ardour { };
 
+  ardour5 = callPackage ../applications/audio/ardour/ardour5.nix { };
```
Comment (veprbl) [inline pkgs/top-level/all-packages.nix:18637]:
```suggestion
  ardour_5 = lowPrio (callPackage ../applications/audio/ardour/5.nix { });
```

File: pkgs/applications/audio/ardour/default.nix [hunk 7]
```diff
@@ -54,25 +61,25 @@ let
   # "git describe" when _not_ on an annotated tag(!): MAJOR.MINOR-REV-HASH.
 
   # Version to build.
-  tag = "5.12";
+  tag = "6.0";
 in stdenv.mkDerivation rec {
   name = "ardour-${tag}";
```
Comment (veprbl) [inline pkgs/applications/audio/ardour/default.nix:66]:
```suggestion
  pname = "ardour";
  version = "6.0";
```

File: pkgs/applications/audio/ardour/default.nix [hunk 8]
```diff
@@ -54,25 +61,25 @@ let
...
 
   src = fetchgit {
     url = "git://git.ardour.org/ardour/ardour.git";
-    rev = "ae0dcdc0c5d13483271065c360e378202d20170a";
-    sha256 = "0mla5lm51ryikc2rrk53max2m7a5ds6i1ai921l2h95wrha45nkr";
+    rev = "360c81b815efe1e7b117ad6f96f8493f675d51bf";
```
Comment (veprbl) [inline pkgs/applications/audio/ardour/default.nix:70]:
```suggestion
    rev = version;
```

File: pkgs/applications/audio/ardour/default.nix (hunk 7 above)
Comment (magnetophon) [reply pkgs/applications/audio/ardour/default.nix:66]: @veprbl Thanks for the review, but like this it replaces ardour5 when installing 6, which what I wanted to avoid, see the commit message about the new latency compensation.

**Comment by prusnak** [comment]
I made a PR earlier here: https://github.com/NixOS/nixpkgs/pull/87274 - please make sure all of the changes are included in your PR and then I will close mine in favor of this one.

Also it's always a good idea to check for other PRs before starting your own to avoid duplicate work.

File: pkgs/applications/audio/ardour/default.nix (hunk 2 above)
Comment (veprbl) [reply pkgs/applications/audio/ardour/default.nix:151]: I guess the reason why definition usually work for waf is because they pick up "gcc" as the entry point: https://gitlab.com/ita1024/waf/-/blob/2e9954e08736e6cfa2aa721073f11f69c9bd5203/waflib/Tools/gas.py#L14
My guesses for what can be special with nixpkgs is
1. Something sets `AS=as`
2. cc-wrapper does something special (I can't see it looking at the code, though)

**Comment by magnetophon** [comment]
@prusnak Oh,I thought I *had* checked for other PRs, sorry!
I incorporated most of your changes, but yours has the same issue as @veprbl 's: it replaces ardour5 when installing 6, which what I wanted to avoid, see the commit message about the new latency compensation.

File: pkgs/applications/audio/ardour/default.nix (hunk 7 above)
Comment (veprbl) [reply pkgs/applications/audio/ardour/default.nix:66]: If you are talking about nix-env, for your current PR it will parse `name` to extract the name of the software, it will find "ardour" and will do the same replacement. If you don't want this to happen, you need to set `pname = "ardour6"` for the new version and/or `pname = "ardour5"` for the old one. In any case, use of `name` attribute is mostly deprecated.

File: pkgs/top-level/all-packages.nix (hunk 6 above)
Comment (magnetophon) [reply pkgs/top-level/all-packages.nix:18637]: OK, done, thanks.

File: pkgs/applications/audio/ardour/default.nix (hunk 7 above)
Comment (magnetophon) [reply pkgs/applications/audio/ardour/default.nix:66]: done

File: pkgs/applications/audio/ardour/default.nix (hunk 8 above)
Comment (magnetophon) [reply pkgs/applications/audio/ardour/default.nix:70]: done

**Comment by veprbl** [comment]
@magnetophon Forgot to push the updated commits?

**Comment by magnetophon** [comment]
@veprbl Ah, yes, something went wrong.

File: pkgs/applications/audio/ardour/5.nix [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+, libusb1
+, fluidsynth_1
+, hidapi
+, libltc
+, qm-dsp
+, libxml2
+, lilv
+, lrdf
+, lv2
+, makeWrapper
```
Comment (prusnak) [inline pkgs/applications/audio/ardour/5.nix:30]: `makeWrapper` is not used, please remove

File: pkgs/applications/audio/ardour/5.nix [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+
+  src = fetchgit {
+    url = "git://git.ardour.org/ardour/ardour.git";
+    rev = "ae0dcdc0c5d13483271065c360e378202d20170a";
+    sha256 = "0mla5lm51ryikc2rrk53max2m7a5ds6i1ai921l2h95wrha45nkr";
+  };
+
+  nativeBuildInputs = [
+    wafHook
+    makeWrapper
```
Comment (prusnak) [inline pkgs/applications/audio/ardour/5.nix:70]: `makeWrapper` is not used, please remove

File: pkgs/applications/audio/ardour/default.nix [hunk 11]
```diff
@@ -5,74 +5,79 @@
...
-, libltc
-, qm-dsp
+, libuv
+, libwebsockets
 , libxml2
+, libxslt
 , lilv
 , lrdf
 , lv2
 , makeWrapper
```
Comment (prusnak) [inline pkgs/applications/audio/ardour/default.nix:39]: `makeWrapper` is not used, please remove

File: pkgs/applications/audio/ardour/default.nix [hunk 12]
```diff
@@ -5,74 +5,79 @@
...
+
   nativeBuildInputs = [
-    wafHook
-    makeWrapper
-    pkg-config
-    itstool
     doxygen
     graphviz # for dot
+    itstool
+    makeWrapper
```
Comment (prusnak) [inline pkgs/applications/audio/ardour/default.nix:76]: `makeWrapper` is not used, please remove

File: pkgs/applications/audio/ardour/default.nix [hunk 13]
```diff
@@ -85,63 +90,71 @@ in stdenv.mkDerivation rec {
...
-    libarchive
   ];
 
   wafConfigureFlags = [
     "--optimize"
     "--docs"
     "--use-external-libs"
     "--freedesktop"
-    "--with-backends=jack,alsa,dummy"
+    "--cxx11"
```
Comment (prusnak) [inline pkgs/applications/audio/ardour/default.nix:136]: maybe add `"--with-backends=jack,alsa,dummy,pulseaudio"` ?

File: pkgs/applications/audio/ardour/default.nix [hunk 14]
```diff
@@ -5,74 +5,79 @@
...
 , lilv
 , lrdf
 , lv2
 , makeWrapper
+, pango
 , perl
-, pkg-config
-, itstool
-, python2
+, pkgconfig
```
Comment (prusnak) [inline pkgs/applications/audio/ardour/default.nix:42]:
```suggestion
, pkg-config
```

File: pkgs/applications/audio/ardour/default.nix [hunk 15]
```diff
@@ -5,74 +5,79 @@
...
-    makeWrapper
-    pkg-config
-    itstool
     doxygen
     graphviz # for dot
+    itstool
+    makeWrapper
     perl
-    python2
+    pkgconfig
```
Comment (prusnak) [inline pkgs/applications/audio/ardour/default.nix:78]:
```suggestion
    pkg-config
```

**Review by prusnak — CHANGES_REQUESTED** [review]

File: pkgs/applications/audio/ardour/default.nix [hunk 16]
```diff
@@ -85,63 +90,71 @@ in stdenv.mkDerivation rec {
...
-    qm-dsp
+    libuv
+    libwebsockets
     libxml2
+    libxslt
     lilv
     lrdf
     lv2
     pango
+    perl
```
Comment (prusnak) [inline pkgs/applications/audio/ardour/default.nix:118]: Why is `perl` needed?

File: pkgs/applications/audio/ardour/default.nix [hunk 17]
```diff
@@ -85,63 +90,71 @@ in stdenv.mkDerivation rec {
...
       -t "$out/share/mime/packages"
-    install -vDm 644 "build/gtk2_ardour/ardour5.desktop" \
+    install -vDm 644 "build/gtk2_ardour/ardour6.desktop" \
       -t "$out/share/applications"
     for size in 16 22 32 48 256 512; do
       install -vDm 644 "gtk2_ardour/resources/Ardour-icon_''${size}px.png" \
-        "$out/share/icons/hicolor/''${size}x''${size}/apps/ardour5.png"
+        "$out/share/icons/hicolor/''${size}x''${size}/apps/ardour6.png"
     done
-    install -vDm 644 "ardour.1"* -t "$out/share/man/man1"
```
Comment (prusnak) [inline pkgs/applications/audio/ardour/default.nix:144]: Why did you remove the manpages?

File: pkgs/applications/audio/ardour/default.nix (hunk 2 above)
Comment (magnetophon) [reply pkgs/applications/audio/ardour/default.nix]: While it would be nice to find the root cause, I think for now the patch provided by @prusnak is a good solution.

File: pkgs/applications/audio/ardour/ardour5.nix (hunk 5 above)
Comment (magnetophon) [reply pkgs/applications/audio/ardour/ardour5.nix]: Seems like at least the freedesktop file is not supposed to be installed by the wscript:
https://github.com/Ardour/ardour/blob/58a5e256a9df97903b59f53f43aef6e4a52febcc/wscript#L774

File: pkgs/applications/audio/ardour/5.nix (hunk 9 above)
Comment (magnetophon) [reply pkgs/applications/audio/ardour/5.nix]: Done, thanks foor the watchful eye!

File: pkgs/applications/audio/ardour/default.nix (hunk 16 above)
Comment (magnetophon) [reply pkgs/applications/audio/ardour/default.nix:117]: It's used in a few tools and in the build iiuc:  https://github.com/Ardour/ardour/search?q=perl&unscoped_q=perl

**Review by prusnak — APPROVED** [review]

File: pkgs/applications/audio/ardour/default.nix [hunk 18]
```diff
@@ -4,146 +4,165 @@
...
       -t "$out/share/applications"
     for size in 16 22 32 48 256 512; do
       install -vDm 644 "gtk2_ardour/resources/Ardour-icon_''${size}px.png" \
-        "$out/share/icons/hicolor/''${size}x''${size}/apps/ardour5.png"
+        "$out/share/icons/hicolor/''${size}x''${size}/apps/ardour6.png"
     done
     install -vDm 644 "ardour.1"* -t "$out/share/man/man1"
   '';
 
+  NIX_LDFLAGS = "-lpthread";
```
Comment (veprbl) [inline pkgs/applications/audio/ardour/default.nix:164]:
```suggestion
  LINKFLAGS = "-lpthread";
```
see rationale #79303

File: pkgs/applications/audio/ardour/default.nix [hunk 19]
```diff
@@ -4,146 +4,165 @@
...
   src = fetchgit {
     url = "git://git.ardour.org/ardour/ardour.git";
-    rev = "ae0dcdc0c5d13483271065c360e378202d20170a";
-    sha256 = "0mla5lm51ryikc2rrk53max2m7a5ds6i1ai921l2h95wrha45nkr";
+    rev = version;
+    sha256 = "162jd96zahl05fdmjwvpdfjxbhd6ifbav6xqa0vv6rsdl4zk395q";
   };
 
+  patches = [
+    # weird "as" issue: https://tracker.ardour.org/view.php?id=8096
```
Comment (veprbl) [inline pkgs/applications/audio/ardour/default.nix:67]:
```suggestion
    # AS=as in the environment causes build failure https://tracker.ardour.org/view.php?id=8096
```
"AS" set by binutils-wrapper:
https://github.com/NixOS/nixpkgs/blob/023e11ae3407c184eff3463eccc4460d5682a9d1/pkgs/build-support/bintools-wrapper/setup-hook.sh#L57-L66

see similar bugs in other software:
https://bugzilla.mozilla.org/show_bug.cgi?id=1497286
https://bugzilla.mozilla.org/show_bug.cgi?id=1497286

**Review by veprbl — APPROVED** [review]
LGTM. Just a few minor comments.

**Comment by magnetophon** [comment]
@veprbl Thanks, all done.

File: pkgs/applications/audio/ardour/default.nix (hunk 19 above)
Comment (magnetophon) [reply pkgs/applications/audio/ardour/default.nix]: @x42 Does the above comment enlighten you as to why the ``as`` issue happens on NixOS and whether the fix should be upstreamed?

**Comment by magnetophon** [comment]
Something went wrong here, I'm trying to fix it!
